### PR TITLE
Last fixes for 0.3.3 pre-release

### PR DESCRIPTION
**Pull request by Zematus**
These are the last few fixes and changes I'm going to make for version 0.3.3. Here's a list: 
- Tried fixing the issues with negative event trigger dates. I couldn't find a root cause but at least I implemented some handling code to prevent the game from crashing. I pushed the related issue (#118) toward 0.3.4. But hopefully I can find a fix before that (or the issue becomes a non-issue)
- Fixed a general game slowdown (#144) due to the agriculture knowledge having an expected progress of 0 in multiple cells.
- Fixed a not so critical issue (#119) that nevertheless generated a lot of warning messages that cluttered the log. This is a legitimate fix though which might prevent more severe issues down the road.
- Cleaned a lot of old DEBUG code (or turned it into release code)
- Reformatted some old code
- Updated game version and title to make them final for 0.3.3

### CODE REVIEW

**Comment by Zematus** [comment]
I ended up adding a last minute commit which updates the Unity3D editor version and the custom game icon which now works!!

Also, I replaced an old debug mode perf statistic which no longer seemed to work, with a new, more useful, perf stat

File: Assets/Scripts/WorldEngine/Cultures/CellCulture.cs [hunk 1]
```diff
@@ -687,24 +687,25 @@ public float MinimumKnowledgeProgressLevel()
...
+            //                        "");
+
+            //                    Manager.RegisterDebugEvent("DebugMessage", debugMessage);
+            //                }
+            //            }
+            //#endif
 
-            float level = knowledge.CalculateExpectedProgressLevel();
+            // if progress level equals 0 that means the knowledge can't really progress. So we ignore it
+            float level = (knowledge.ProgressLevel > 0) ? knowledge.CalculateExpectedProgressLevel() : 1;
```
Comment (nsmonagh) [inline Assets/Scripts/WorldEngine/Cultures/CellCulture.cs]: So the entire fix was just to add this statement? That's pretty nice.
How do knowledges get stuck at 0 progress level?

File: Assets/Scripts/2D/GuiManagerScript.cs [hunk 2]
```diff
@@ -344,7 +344,8 @@ void Start()
             //GenerateWorld(false, 1645709120);
             //GenerateWorld(false, 888101979);
             //GenerateWorld(false, 6353535);
-            GenerateWorld(false, 1137426545);
+            //GenerateWorld(false, 1137426545);
+            GenerateWorld(false, 1277025723);
```
Comment (nsmonagh) [inline Assets/Scripts/2D/GuiManagerScript.cs]: What is with all the GenerateWorld calls here? I've seen it before but I assume its for debugging?

File: Assets/Scripts/WorldEngine/World.cs [hunk 3]
```diff
@@ -1384,15 +1384,15 @@ public long Update()
...
-        UpdateFactions();
+        RemoveFactions();
 
         Profiler.EndSample();
 
-        Profiler.BeginSample("RemoveFactions");
+        Profiler.BeginSample("UpdateFactions");
 
-        RemoveFactions();
+        UpdateFactions();
```
Comment (nsmonagh) [inline Assets/Scripts/WorldEngine/World.cs]: I'm guessing this was the fix for #119?

File: Assets/Tools/Scripts/MathUtility.cs [hunk 4]
```diff
@@ -129,7 +129,7 @@ public static Vector2 GetSphereProjection(float x, float y, float z, float alpha
     public static float RoundToSixDecimals(float value)
     {
 #if DEBUG
-        if ((value < 0) || (value > 1))
+        if (!value.IsInsideRange(0,1))
```
Comment (nsmonagh) [inline Assets/Tools/Scripts/MathUtility.cs]: Good addition.

File: Assets/Scripts/WorldEngine/World.cs [hunk 5]
```diff
@@ -1750,24 +1779,9 @@ public void AddFactionToUpdate(Faction faction)
...
-                || callingMethod.Contains("SetToUpdate"))
-            {
-                method = stackTrace.GetFrame(frame).GetMethod();
-                callingMethod = method.Name;
-
-                frame++;
-            }
-
-            string callingClass = method.DeclaringType.ToString();
-
```
Comment (nsmonagh) [inline Assets/Scripts/WorldEngine/World.cs]: Why was this removed?

File: Assets/Scripts/WorldEngine/Cultures/CellCulture.cs (hunk 1 above)
Comment (Zematus) [reply Assets/Scripts/WorldEngine/Cultures/CellCulture.cs]: That's a deeper issue that I need to investigate further. One possible cause is that groups acquire a knowledge from a neighbor but do not have the necessary conditions to sustain it so the knowledge drops to 0. But not all knowledges have a removal mechanism so they linger at 0 forever. This is bad because any active knowledge requires to be updated, which means wasted CPU cycles for knowledges that have no way to progress.

File: Assets/Scripts/2D/GuiManagerScript.cs (hunk 2 above)
Comment (Zematus) [reply Assets/Scripts/2D/GuiManagerScript.cs]: Yeah, it's just a collection of interesting worlds that I use to play test during development. A few I added to test specific bugs that have been fixed since. I should do a cleanup though. but I'm a code hoarder.

File: Assets/Scripts/WorldEngine/World.cs (hunk 3 above)
Comment (Zematus) [reply Assets/Scripts/WorldEngine/World.cs]: Yes, turns out I never remove factions during faction updates, so the fix was simply to do faction updates after faction removals. I probably could do the same with polities.

File: Assets/Scripts/WorldEngine/World.cs (hunk 5 above)
Comment (Zematus) [reply Assets/Scripts/WorldEngine/World.cs]: System reflection ops are very CPU and memory intensive. So it's better to remove them if not needed. In most cases it's just simpler to do a ToString() on a stack trace object instead.

**Comment by Zematus** [comment]
Thanks for the review!
